### PR TITLE
replace QDateTime::fromTime_t with QDateTime::fromSecsSinceEpoch

### PR DESCRIPTION
**Pull request by fanquake**
`QDateTime::fromTime_t` has [been obsoleted](https://doc.qt.io/qt-5.12/qdatetime-obsolete.html#fromTime_t) in favour of [`QDateTime::fromSecsSinceEpoch`](https://doc.qt.io/qt-5.12/qdatetime.html#fromSecsSinceEpoch), which is available from Qt 5.8+.

### CODE REVIEW

**Comment by hebasto** [comment]
Concept ACK.

Interesting that there are no `[-Wdeprecated-declarations]` warnings in the build log (Fedora 34 + Qt 5.15.2).

File: src/qt/guiutil.cpp [hunk 1]
```diff
@@ -81,7 +81,7 @@ QString dateTimeStr(const QDateTime &date)
 
 QString dateTimeStr(qint64 nTime)
 {
-    return dateTimeStr(QDateTime::fromTime_t((qint32)nTime));
+    return dateTimeStr(QDateTime::fromSecsSinceEpoch((qint32)nTime));
```
Comment (MarcoFalke) [inline src/qt/guiutil.cpp:84]:
```suggestion
    return dateTimeStr(QDateTime::fromSecsSinceEpoch(nTime));
```

File: src/qt/transactiontablemodel.cpp [hunk 2]
```diff
@@ -610,7 +610,7 @@ QVariant TransactionTableModel::data(const QModelIndex &index, int role) const
     case TypeRole:
         return rec->type;
     case DateRole:
-        return QDateTime::fromTime_t(static_cast<uint>(rec->time));
+        return QDateTime::fromSecsSinceEpoch(static_cast<uint>(rec->time));
```
Comment (MarcoFalke) [inline src/qt/transactiontablemodel.cpp:613]:
```suggestion
        return QDateTime::fromSecsSinceEpoch(qint64{rec->time});
```

File: src/qt/transactiontablemodel.cpp [hunk 3]
```diff
@@ -630,7 +630,7 @@ QVariant TransactionTableModel::data(const QModelIndex &index, int role) const
     case TxPlainTextRole:
         {
             QString details;
-            QDateTime date = QDateTime::fromTime_t(static_cast<uint>(rec->time));
+            QDateTime date = QDateTime::fromSecsSinceEpoch(static_cast<uint>(rec->time));
```
Comment (MarcoFalke) [inline src/qt/transactiontablemodel.cpp:633]: same

**Review by MarcoFalke — COMMENTED** [review]
You are passing the wrong type. Otherwise Concept ACK

File: src/qt/transactionfilterproxy.cpp [hunk 4]
```diff
@@ -10,9 +10,9 @@
 #include <cstdlib>
 
 // Earliest date that can be represented (far in the past)
-const QDateTime TransactionFilterProxy::MIN_DATE = QDateTime::fromTime_t(0);
+const QDateTime TransactionFilterProxy::MIN_DATE = QDateTime::fromSecsSinceEpoch(0);
 // Last date that can be represented (far in the future)
-const QDateTime TransactionFilterProxy::MAX_DATE = QDateTime::fromTime_t(0xFFFFFFFF);
+const QDateTime TransactionFilterProxy::MAX_DATE = QDateTime::fromSecsSinceEpoch(0xFFFFFFFF);
```
Comment (hebasto) [inline src/qt/transactionfilterproxy.cpp:15]: If @MarcoFalke's suggestions will be addressed, the max `qint64` could be also fixed.

**Comment by promag** [comment]
Concept ACK. No longer scripted-diff.

File: src/qt/transactionfilterproxy.cpp (hunk 4 above)
Comment (MarcoFalke) [inline src/qt/transactionfilterproxy.cpp:15]:
```suggestion
const QDateTime TransactionFilterProxy::MAX_DATE = QDateTime::fromSecsSinceEpoch(std::numeric_limits<qint64>::max());
```

not sure if this compiles

**Review by MarcoFalke — APPROVED** [review]
ACK

File: src/qt/transactionfilterproxy.cpp (hunk 4 above)
Comment (promag) [reply src/qt/transactionfilterproxy.cpp:15]: I'm proposing to drop these in #354.

Comment (hebasto) [reply src/qt/transactionfilterproxy.cpp:15]: > not sure if this compiles

It does.

Also the `<limits>` header could be included:

``` diff
--- a/src/qt/transactionfilterproxy.cpp
+++ b/src/qt/transactionfilterproxy.cpp
@@ -8,11 +8,12 @@
 #include <qt/transactionrecord.h>
 
 #include <cstdlib>
+#include <limits>
 
 // Earliest date that can be represented (far in the past)
 const QDateTime TransactionFilterProxy::MIN_DATE = QDateTime::fromSecsSinceEpoch(0);
 // Last date that can be represented (far in the future)
-const QDateTime TransactionFilterProxy::MAX_DATE = QDateTime::fromSecsSinceEpoch(0xFFFFFFFF);
+const QDateTime TransactionFilterProxy::MAX_DATE = QDateTime::fromSecsSinceEpoch(std::numeric_limits<qint64>::max());
 
 TransactionFilterProxy::TransactionFilterProxy(QObject *parent) :
     QSortFilterProxyModel(parent),
```

File: src/qt/transactiontablemodel.cpp [hunk 5]
```diff
@@ -610,7 +610,7 @@ QVariant TransactionTableModel::data(const QModelIndex &index, int role) const
     case TypeRole:
         return rec->type;
     case DateRole:
-        return QDateTime::fromTime_t(static_cast<uint>(rec->time));
+        return QDateTime::fromSecsSinceEpoch(qint64{rec->time});
```
Comment (hebasto) [inline src/qt/transactiontablemodel.cpp:613]: `rec->time` is of type `qint64`:
https://github.com/bitcoin-core/gui/blob/e033ca13794699cf4744e71647db75c583a9a600/src/qt/transactionrecord.h#L116

File: src/qt/transactiontablemodel.cpp [hunk 6]
```diff
@@ -630,7 +630,7 @@ QVariant TransactionTableModel::data(const QModelIndex &index, int role) const
     case TxPlainTextRole:
         {
             QString details;
-            QDateTime date = QDateTime::fromTime_t(static_cast<uint>(rec->time));
+            QDateTime date = QDateTime::fromSecsSinceEpoch(qint64{rec->time});
```
Comment (hebasto) [inline src/qt/transactiontablemodel.cpp:633]: `rec->time` is of type `qint64`:
https://github.com/bitcoin-core/gui/blob/e033ca13794699cf4744e71647db75c583a9a600/src/qt/transactionrecord.h#L116

**Review by hebasto — CHANGES_REQUESTED** [review]

File: src/qt/transactiontablemodel.cpp (hunk 2 above)
Comment (fanquake) [reply src/qt/transactiontablemodel.cpp]: Reverted per https://github.com/bitcoin-core/gui/pull/349#discussion_r646104519.

File: src/qt/transactiontablemodel.cpp (hunk 3 above)
Comment (fanquake) [reply src/qt/transactiontablemodel.cpp]: [same](https://github.com/bitcoin-core/gui/pull/349#discussion_r646276429)

**Comment by MarcoFalke** [comment]
cr ACK 4f5e9feb64400948223ae153b74eb41811f53cff

**Comment by fanquake** [comment]
Added one additional commit to replace `QDateTime::toTime_t` with `QDateTime::toSecsSinceEpoch`.

**Review by hebasto — APPROVED** [review]
ACK bd533e785903af0ba04cb0511c907b98008b056c,

```
$ git grep fromTime_t | wc -l
0
$ git grep toTime_t | wc -l
0
```

Maybe update the OP to mention the second commit changes?

File: src/qt/recentrequeststablemodel.cpp [hunk 7]
```diff
@@ -234,7 +234,7 @@ bool RecentRequestEntryLessThan::operator()(const RecentRequestEntry& left, cons
     switch(column)
     {
     case RecentRequestsTableModel::Date:
-        return pLeft->date.toTime_t() < pRight->date.toTime_t();
+        return pLeft->date.toSecsSinceEpoch() < pRight->date.toSecsSinceEpoch();
```
Comment (Talkless) [inline src/qt/recentrequeststablemodel.cpp:237]: Out of scope, but I wonder why `bool QDateTime::operator<` is not used directly, without `toTimeT()/toSecsSinceEpoch()` ?

File: src/qt/transactionfilterproxy.cpp [hunk 8]
```diff
@@ -8,11 +8,12 @@
...
 
 #include <cstdlib>
+#include <limits>
 
 // Earliest date that can be represented (far in the past)
-const QDateTime TransactionFilterProxy::MIN_DATE = QDateTime::fromTime_t(0);
+const QDateTime TransactionFilterProxy::MIN_DATE = QDateTime::fromSecsSinceEpoch(0);
 // Last date that can be represented (far in the future)
-const QDateTime TransactionFilterProxy::MAX_DATE = QDateTime::fromTime_t(0xFFFFFFFF);
+const QDateTime TransactionFilterProxy::MAX_DATE = QDateTime::fromSecsSinceEpoch(std::numeric_limits<qint64>::max());
```
Comment (Talkless) [inline src/qt/transactionfilterproxy.cpp:16]: From Qt documentation [0]:

> Note that there are possible values for secs that lie outside the valid range of QDateTime, both negative and positive. The behavior of this function is undefined for those values.

Anyone knows what are these "ouside" values, and is this `::max()` OK here then?

[0] https://doc.qt.io/qt-5/qdatetime.html#fromSecsSinceEpoch

**Review by Talkless — COMMENTED** [review]
Concept ACK

File: src/qt/transactionfilterproxy.cpp (hunk 8 above)
Comment (Talkless) [reply src/qt/transactionfilterproxy.cpp:16]: I guess we have to wait for https://github.com/bitcoin-core/gui/pull/354 to be merged, these lines will go away.

**Comment by MarcoFalke** [comment]
cr ACK bfcde3e87231d90bdacd3a8154e2be16a52508c7

File: src/qt/recentrequeststablemodel.h [hunk 9]
```diff
@@ -26,9 +26,9 @@ class RecentRequestEntry
 
     SERIALIZE_METHODS(RecentRequestEntry, obj) {
         unsigned int date_timet;
```
Comment (MarcoFalke) [inline src/qt/recentrequeststablemodel.h:28]: unrelated: This is still the wrong type

File: src/qt/transactionfilterproxy.cpp [hunk 10]
```diff
@@ -8,6 +8,7 @@
 #include <qt/transactionrecord.h>
 
 #include <cstdlib>
+#include <limits>
```
Comment (hebasto) [inline src/qt/transactionfilterproxy.cpp:11]: This change is unrelated after the recent rebase and could be dropped.

**Review by hebasto — COMMENTED** [review]
Approach ACK bfcde3e87231d90bdacd3a8154e2be16a52508c7.

File: src/qt/transactionfilterproxy.cpp (hunk 10 above)
Comment (fanquake) [reply src/qt/transactionfilterproxy.cpp]: Replaced it with the headers that are actually used.

File: src/qt/recentrequeststablemodel.h (hunk 9 above)
Comment (fanquake) [reply src/qt/recentrequeststablemodel.h]: Changed the type here.

**Comment by hebasto** [comment]
```
./serialize.h:676:6: error: member reference base type 'long long' is not a structure or union
    a.Unserialize(is);
    ~^~~~~~~~~~~~
```

File: src/qt/recentrequeststablemodel.h [hunk 11]
```diff
@@ -25,10 +27,10 @@ class RecentRequestEntry
     SendCoinsRecipient recipient;
 
     SERIALIZE_METHODS(RecentRequestEntry, obj) {
-        unsigned int date_timet;
-        SER_WRITE(obj, date_timet = obj.date.toTime_t());
+        qint64 date_timet;
```
Comment (MarcoFalke) [inline src/qt/recentrequeststablemodel.h:30]: I don't think this can be changed without breaking wallet.dat

Comment (fanquake) [reply src/qt/recentrequeststablemodel.h:30]: I'm going to revert this. We can leave it as unrelated.

**Review by hebasto — APPROVED** [review]
ACK 3ae503c95bccc4899fc881c603d618a824078e7b
